src/vhip/mlmodel/gene_features.py

```diff
@@ -310,3 +310,45 @@ def amino_acid_frequency(
         if hasattr(self, "aa_dict"):
             total = sum(self.aa_dict.values())
             self.aa_frq = {k: (v / total) for k, v in self.aa_dict.items()}
+
+    def RSCU(
+        self, threshold_imprecise: float = 0.0, threshold_skipped_genes: float = 0.5
+    ) -> None:
+        """Calculate the relative synonymous codon usage (RSCU) of each codon in entire genome.
+
+        Args:
+            threshold_imprecise (float): Percentage of imprecise (non-ATGC) codons tolerated in a single gene (default 0.0 or 0%)
+            threshold_skipped_genes (float): Tolerated percentage of valid (codon length divisible) genes in GeneSet that have more than threshold_imprecise codons (default 0.5 or 50%)
+        Definitions:
+            Synonymous codons: codons that encode the same amino acid
+            RSCU_dict: codon count / expected frequency (given assumption of equally used synonymous codons)
+        Populates the following class attributes:
+            self.RSCU (str: float): RSCU of each codon across all genes in the GeneSet.
+        If not populated previously by running codon_counts() or codon_frequency():
+            self.codon_dict (str: int): Counts of each unique codon across all genes in the GeneSet.
+            self.imprecise_codons (int): Total number of imprecise codons found in the GeneSet.
+            self.skipped_imprecise_genes (List[str]): IDs of genes in the GeneSet that have more than threshold_imprecise codons.
+        """
+        self.RSCU_dict: dict[str, float] = dict.fromkeys(CODON_LIST, 0.0)
+
+        if not hasattr(self, "codon_dict"):
+            # If aggregate codon counts have not already been calculated, runs codon_counts()
+            self.codon_counts(
+                threshold_imprecise=threshold_imprecise,
+                threshold_skipped_genes=threshold_skipped_genes,
+            )
+
+        if hasattr(self, "codon_dict"):
+            for codon, count in self.codon_dict.items():
+                if count != 0.0:
```

Comment:
why are you checking if the count is 0? and if it is 0, then what should happen?

Reply:
So, the self.RSCU_dict is initialized with 0.0 for all codons. If a codon is not there, it should have an RSCU of 0 because the numerator (count) is 0 - and if the count is 0, the initialized value doesn't need to change. I thought having this check might be optimal, letting the algorithm avoid performing any calculations that are not needed - not sure, this might help in a very minute way because I'm not sure how many of the 64 codons are typically not found in a given gene. What do you think?